Runtime/JSError.cs

```diff
@@ -0,0 +1,302 @@
+
+        // A pending JS exception takes precedence over any internal error status.
+        if (IsExceptionPending())
+        {
+            _message = null;
+            _errorRef = CreateErrorReference(GetAndClearLastException());
+            return;
+        }
+
+        var messageBuilder = new StringBuilder();
+        if (!string.IsNullOrEmpty(errorInfo.Message))
+        {
+            messageBuilder.Append(errorInfo.Message);
+        }
+
+        if (!string.IsNullOrEmpty(message))
+        {
+            if (messageBuilder.Length != 0)
+            {
+                messageBuilder.Append(": ");
+            }
+            messageBuilder.Append(message);
+        }
+
+        if (messageBuilder.Length == 0)
+        {
+            messageBuilder.Append(DefaultMessage);
+        }
+
+        _message = messageBuilder.ToString();
+        _errorRef = CreateErrorReference(errorInfo.Status switch
+        {
+            napi_status.napi_ok => errorType switch
+            {
+                JSErrorType.TypeError => JSValue.CreateTypeError(null, (JSValue)_message),
+                JSErrorType.RangeError => JSValue.CreateRangeError(null, (JSValue)_message),
+                JSErrorType.SyntaxError => JSValue.CreateSyntaxError(null, (JSValue)_message),
+                _ => JSValue.CreateError(null, (JSValue)_message),
+            },
+            napi_status.napi_object_expected or
+            napi_status.napi_string_expected or
+            napi_status.napi_boolean_expected or
+            napi_status.napi_number_expected
+              => JSValue.CreateTypeError(null, (JSValue)_message),
+            _ => JSValue.CreateError(null, (JSValue)_message),
+        });
+    }
+
+    public JSError(JSValue? error)
+    {
+        if (error is null)
+            return;
+
+        _errorRef = CreateErrorReference(error.Value);
+    }
+
+    public JSError(Exception exception)
+    {
+        if (exception is JSException jsException)
+        {
+            _message = jsException.Error._message;
+            _errorRef = jsException.Error._errorRef;
+            return;
+        }
+
+        var tempError = new JSError(exception.Message);
+        _message = tempError._message;
+        _errorRef = tempError._errorRef;
+    }
+
+    public string Message
+    {
+        get
+        {
+            if (_message is null && _errorRef is not null)
+            {
+                try
+                {
+                    _message = (string?)_errorRef.GetValue()?["message"];
+                }
+                catch
+                {
+                    // Catch all errors here, because this method may not throw.
+                }
+            }
+
+            return _message ?? DefaultMessage;
+        }
+    }
+
+    public JSValue? Value
+    {
+        get
+        {
+            JSValue? error = _errorRef?.GetValue();
+            if (error is null)
+                return null;
```

Comment:
I think it would be more convenient & consistent if the `JSError.Value` property is not nullable, then this could return undefined. The `Message` property returns a default non-null value.

Reply:
Good point! Let me change it.